### PR TITLE
Support merging multiple result sets in QueryExecutor

### PR DESCRIPTION
**Pull request by dimas-b**
Contributes to #828 

-----

This PR adds support to `QueryExecutor` for running more than one query (against the same table) while merging the result sets on Document identity columns (C* primary key).

Paging state is extended to allow advancing pages in each of the underlying queries independently, as the iteration over the merged result set progresses.

This include minor changes to the handling of document pagination at the `RawDocument` level. In particular the `hasNext` flag is no longer necessary.

No functional changes are expected in the single query case.

### CODE REVIEW

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 1]
```diff
@@ -57,30 +74,109 @@ public QueryExecutor(DataStore dataStore) {
...
+      }
+    }
+
+    if (pathColumns == null) {
+      throw new IllegalArgumentException("No query provided");
+    }
+
+    Comparator<DocProperty> comparator = Comparator.comparing(DocProperty::comparableKey);
+    for (Column column : pathColumns) {
+      comparator = comparator.thenComparing(p -> p.keyValue(column));
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:123]: why we are including this extra? Isn't enough that the order is per key? Why do we care for the other sub-document props here? Who cares how are the sub-documents of the single document order, that's all on the same partition?

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 2]
```diff
@@ -116,20 +212,37 @@ public QueryExecutor(DataStore dataStore) {
...
+    String id = row.getString("key");
+    Builder<String> docKey = ImmutableList.builder();
+    for (Column c : keyColumns) {
+      docKey.add(Objects.requireNonNull(row.getString(c.name())));
+    }
+
+    return new Accumulator(id, comparator, docKey.build(), property);
+  }
+
+  private Iterable<DocProperty> properties(
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:227]: can you explain this a bit more to me please? also some kind of inline docs would be so helpful to understand what s going on, nobody will ever be able to understand this except the writer :)

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 3]
```diff
@@ -139,39 +252,51 @@ public DataStore getDataStore() {
   private class Accumulator {
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:252]: could be worth to extract this to a dedicated file now.. Stands true for the other private classes as well..

**Review by ivansenic — CHANGES_REQUESTED** [review]
It's very hard to follow the changes in the `QueryExecutor` without any inline comments or PR description. I did a first look on the whole code, but I would need some guidance to finalize the review..

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      } else {
+        result.putInt(-1); // no state
+      }
+    }
+
+    result.flip();
+    return result;
+  }
+
+  public static CombinedPagingState deserialize(int expectedSize, ByteBuffer data) {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java:69]: Yes, it probably makes sense to make it `static` too... will do.

[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java:69]: The single element case is not really a workaround, but more of a backward compatibility logic for supporting rolling upgrades where single query paging state may be generated by older code, but interpreted by newer code.

[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java:69]: `expectedSize` is a sanity check. Ideally it should be something like a hash of the whole query plan, but doing that seemed like an overkill at this point.

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+              "Invalid paging state: expected element count: %d, actual: %d", expectedSize, count));
+    }
+
+    ArrayList<ByteBuffer> nested = new ArrayList<>(count);
+    while (count-- > 0) {
+      int size = data.getInt();
+
+      ByteBuffer element;
+      if (size >= 0) {
+        element = data.slice();
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java:106]: sure

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 6]
```diff
@@ -57,30 +74,109 @@ public QueryExecutor(DataStore dataStore) {
...
+  private List<Column> idColumns(int keyDepth, List<BoundQuery> queries) {
+    List<Column> idColumns = null;
+    for (BoundQuery query : queries) {
+      BuiltSelect select = (BuiltSelect) query.source().query();
+      if (keyDepth < 1 || keyDepth > select.table().primaryKeyColumns().size()) {
+        throw new IllegalArgumentException("Invalid document identity depth: " + keyDepth);
+      }
+
+      if (idColumns == null) {
+        idColumns = select.table().primaryKeyColumns().subList(0, keyDepth);
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:138]: All queries must reference the same table and have the same selectors. I'll add a check for that. Thanks for flagging this!

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 7]
```diff
@@ -57,30 +74,109 @@ public QueryExecutor(DataStore dataStore) {
...
+      int pageSize,
+      CombinedPagingState pagingState,
+      ExecutionContext context) {
+    List<Flowable<DocProperty>> flows = new ArrayList<>(queries.size());
+    int idx = 0;
+    for (BoundQuery query : queries) {
+      final int finalIdx = idx++;
+      ByteBuffer queryPagingState = pagingState.nested().get(finalIdx);
+      if (queryPagingState != null) {
+        queryPagingState = queryPagingState.slice();
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:161]: good point.

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 1 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:123]: We need strong order to be able to identify duplicate rows without loading all of the data into memory. Currently we assume all queries have the same sorting order (according to `pNNN` clustering columns). Then, we can merge the result sets by comparing only the top-level elements.

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 2 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:227]: I'll add comments

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 3 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:252]: I'm not sure about that. The logic encapsulated into `Accumulator` is tightly coupled to the way our Rx pipeline is constructed, in particular to the use of the `scan` operator. Let's keep it private for now. I'd be open to factoring it out to a top-level class later, though.

I will move other general-purpose helper classes to top level.

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java:116]: I made the `Paginator`'s buffer read only instead in fe2d63fd36a102c70a867ac58b63bb72d447a6cb, WDYT? A slices or a R/O buffer will be read-only too.

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java]: done in fe2d63fd36a102c70a867ac58b63bb72d447a6cb

[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java]: comments added in fe2d63fd36a102c70a867ac58b63bb72d447a6cb

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  public ByteBuffer serialize() {
+    if (nestedStates.stream().allMatch(Objects::isNull)) {
+      return null;
+    }
+
+    if (nestedStates.size() == 1) {
+      return nestedStates.get(0);
+    }
+
+    int toAllocate = 4; // size of int for element count
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java]: done in fe2d63fd36a102c70a867ac58b63bb72d447a6cb

File: restapi/src/test/java/io/stargate/db/datastore/AlphabeticalOrderPartitionKeyDecorator.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ */
+package io.stargate.db.datastore;
+
+import io.stargate.db.ComparableKey;
+import io.stargate.db.RowDecorator;
+import io.stargate.db.schema.Column;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class AlphabeticalOrderPartitionKeyDecorator implements RowDecorator {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/test/java/io/stargate/db/datastore/AlphabeticalOrderPartitionKeyDecorator.java:28]: Added in 274d75aece0c491c4c367d30c7b246ae6280d036

File: restapi/src/test/java/io/stargate/db/datastore/AlphabeticalOrderPartitionKeyDecorator.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  private final List<Column> partitionKeyColumns;
+
+  public AlphabeticalOrderPartitionKeyDecorator(List<Column> allColumns) {
+    this.partitionKeyColumns =
+        allColumns.stream().filter(Column::isPartitionKey).collect(Collectors.toList());
+  }
+
+  @Override
+  public <T extends Comparable<T>> ComparableKey<T> decoratePartitionKey(Row row) {
+    StringBuilder decorated = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/test/java/io/stargate/db/datastore/AlphabeticalOrderPartitionKeyDecorator.java]: Thanks for the tip! Implemented in 274d75aece0c491c4c367d30c7b246ae6280d036

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class CombinedPagingState {
+
+  private final List<ByteBuffer> nestedStates;
+
+  private CombinedPagingState(List<ByteBuffer> nestedStates) {
+    this.nestedStates = nestedStates;
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java]: I guess this is no longer necessary after the refactoring done in fe2d63fd36a102c70a867ac58b63bb72d447a6cb

File: restapi/src/main/java/io/stargate/web/docsapi/service/PagingStateSupplier.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.web.docsapi.service;
+
+import io.stargate.db.PagingPosition.ResumeMode;
+import java.nio.ByteBuffer;
+
+public abstract class PagingStateSupplier {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/PagingStateSupplier.java]: Added in 27dc6ca381a43d4285718bf58e20192ed7e40c4c

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 3 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:252]: I've refactored all of the paging state handlers and suppliers in 27dc6ca381a43d4285718bf58e20192ed7e40c4c. Please review again.

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivansenic) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java:152]: Perfect

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 1 above)
Comment (ivansenic) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:123]: Hmm very good point.. I did not think about the duplication of the rows from the different queries..

So how is this handled? Where are duplicates dismissed? I guess `orderedMerge` will not do that? Is this something we handle in the Accumulator..

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 13]
```diff
@@ -183,34 +308,52 @@ public RawDocument toDoc() {
...
-      lastResultSet = other.lastResultSet;
+      if (rowComparator == null || lastRow == null) {
+        throw new IllegalStateException("Cannot compare");
+      }
+
+      DocProperty otherRow = other.lastRow;
+
+      if (rowComparator.compare(this.lastRow, otherRow) != 0) {
+        rows.add(otherRow);
+        this.lastRow = otherRow;
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:350]: so is this correct, you'll always update the last row, no matter if the comparator returns `1` or `-1`? How's that?

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 14]
```diff
@@ -219,15 +362,103 @@ private Accumulator combine(Accumulator buffer) {
...
+    public void track(DocProperty row) {
+      states.set(row.queryIndex(), row);
+    }
+
+    public List<PagingStateSupplier> slice() {
+      return new ArrayList<>(states);
+    }
+  }
+
+  /**
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:398]: the comments here really helped :+1:

File: restapi/src/test/java/io/stargate/web/docsapi/service/CombinedPagingStateTest.java [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+          new CombinedPagingState(ImmutableList.of(resumeMode -> null, resumeMode -> null));
+      assertThat(combined.makePagingState(ResumeMode.NEXT_PARTITION)).isNull();
+    }
+
+    @Test
+    void mixed() {
+      CombinedPagingState combined =
+          new CombinedPagingState(
+              ImmutableList.of(resumeMode -> null, resumeMode -> EXHAUSTED_PAGE_STATE));
+      assertThat(combined.makePagingState(ResumeMode.NEXT_PARTITION)).isNotNull();
```
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/CombinedPagingStateTest.java:78]: is this correct considering the tests above and below?

File: restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java [hunk 16]
```diff
@@ -414,4 +422,160 @@ void testSubDocumentsPaged(int pageSize) {
...
+        executor.queryDocs(ImmutableList.of(q1, q2), pageSize, null, context).test();
+    test.assertNoErrors();
+
+    List<RawDocument> docs = test.values();
+
+    assertThat(docs).extracting(RawDocument::id).containsExactly("a", "b", "c", "d");
+    assertThat(docs.get(0).rows()).extracting(r -> r.getString("p0")).contains("x1", "x2");
+    assertThat(docs.get(1).rows()).extracting(r -> r.getString("p0")).contains("y2");
+    assertThat(docs.get(2).rows()).extracting(r -> r.getString("p0")).contains("x2", "x3");
+    assertThat(docs.get(3).rows()).extracting(r -> r.getString("p0")).contains("z1", "z2");
```
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java:463]: :partying_face:

File: restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java [hunk 17]
```diff
@@ -414,4 +422,160 @@ void testSubDocumentsPaged(int pageSize) {
       assertThat(docs).isEmpty();
     }
   }
+
+  @ParameterizedTest
+  @CsvSource({"1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "100"})
+  void testMergeByDocKey(int pageSize) {
```
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java:428]: can we have the same extended depth please?

**Review by ivansenic — COMMENTED** [review]
This is really awesome work.. Well done @dimas-b 

Tests failing though :(

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 1 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:123]: `orderedMerge` will merge source flows placing equal rows next to each other. `Accumulator.combine(...)` later will take only one of a group of equal rows.

File: restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java (hunk 17 above)
Comment (dimas-b) [reply restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java:428]: will add

**Comment by dimas-b** [comment]
@ivansenic : Re: test failures: I had to undo (f0adb4940e9cd06e7504192d0f03ebd0bf213f04) the read-only ByteBuffer change that you previously suggested.

The reason is that C* (3.11 at least) appears to handle read-only byte buffers incorrectly in `FastByteOperations$UnsafeOperations.copy`. It gets the "address" as `0` and tries to copy memory from there, resulting in `SIGSEGV` :(

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java:152]: fe2d63f had to be reverted, unfortunately: https://github.com/stargate/stargate/pull/1091#issuecomment-877358241

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java]: I've refactored this code  and added comments in the last round of updates. Please re-review.

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 18]
```diff
@@ -57,30 +74,109 @@ public QueryExecutor(DataStore dataStore) {
...
         .filter(Accumulator::isComplete)
         .map(Accumulator::toDoc);
   }
 
+  private Comparator<DocProperty> rowComparator(List<BoundQuery> queries) {
+    List<Column> pathColumns = null;
+    for (BoundQuery query : queries) {
+      BuiltSelect select = (BuiltSelect) query.source().query();
+
+      if (pathColumns == null) {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java]: I've refactored this code  and added comments in the last round of updates. Please re-review.

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 2 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:245]: javadoc added

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 13 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java]: We only care about equality at this point.

I hope it's a bit clearer in the refactored code. Please see the latest commits.

File: restapi/src/test/java/io/stargate/web/docsapi/service/CombinedPagingStateTest.java (hunk 15 above)
Comment (dimas-b) [reply restapi/src/test/java/io/stargate/web/docsapi/service/CombinedPagingStateTest.java:78]: I think it is correct. If one of the results sets was never processed (`null`), while the other  is exhausted, we want a non-null paging state so that the user could ask for more data and cause the result set of the former query to be processed.

File: restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java (hunk 17 above)
Comment (dimas-b) [reply restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java:430]: tests added in the new set of commits

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    // The special case for one nested element is mostly to support rolling upgrades where older
+    // nodes may be producing one-elelement paging states, while newer (upgraded) nodes might have
+    // to interpret those paging states.
+    if (expectedSize == 1) {
+      return ImmutableList.of(data);
+    }
+
+    if (data.remaining() < 4) {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java]: thx! fixed in cf003f15866f12f2c303f583a85e18fcc2f931f6

File: restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+  @Override
+  public ByteBuffer makePagingState(ResumeMode resumeMode) {
+    List<ByteBuffer> pagingStateBuffers =
+        pagingState.stream().map(s -> s.makePagingState(resumeMode)).collect(Collectors.toList());
+
+    // An empty paging state means the query was exhausted during previous execution.
+    // A null paging state mean the query has not been executed yet.
+    if (pagingStateBuffers.stream().allMatch(CombinedPagingState::isExhausted)) {
+      // Return a null overall paging state if all queries have been exhausted.
+      return null;
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/CombinedPagingState.java:64]: The trick is that "exhausted" paging states may occurs as a result of a round trip through the HTTP-level paging state, in which case they will be separate buffers, empty, but not identical to `EXHAUSTED_PAGE_STATE`.

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:174]: In the end I think slicing is easier to understand and safer. Paging state buffers may be read by different pieces of code, including C* code and our `QueryExecutor` code (for building the "combined" paging state). Slicing ensures that the buffer pointers we use are independent. Resetting should work, but then we have think about the correctness of the read/reset sequence of operations, which is a bit tricky in this context. I hope slicing does not add any measurable overhead... WDYT?

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 21]
```diff
@@ -57,30 +76,120 @@ public QueryExecutor(DataStore dataStore) {
       int pageSize,
       ByteBuffer pagingState,
       ExecutionContext context) {
+    return queryDocs(keyDepth, ImmutableList.of(query), pageSize, pagingState, context);
+  }
+
+  /**
+   * Runs the provided queries in parallel and merges their result sets according to the selected
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:83]: ohh this is so nice, we needed this kind of detailed explanation what s possible what not

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ivansenic) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:174]: Fine with me

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 22]
```diff
@@ -116,20 +225,42 @@ public QueryExecutor(DataStore dataStore) {
     }
   }
 
-  private Iterable<Accumulator> seeds(
-      BoundQuery query, ResultSet rs, List<Column> keyColumns, ExecutionContext context) {
+  private Accumulator toSeed(
+      DocProperty property, Comparator<DocProperty> comparator, List<Column> keyColumns) {
+    Row row = property.row();
+    String id = row.getString("key");
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:231]: nit: I think we should stop having the column names used by the DocsAPI directly as the literal, wanna use the constant value from the `io.stargate.web.docsapi.service.query.QueryConstants`?

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java [hunk 23]
```diff
@@ -219,15 +389,191 @@ private Accumulator combine(Accumulator buffer) {
...
+
+    private boolean isSelected(Column column) {
+      // An empty selection set means `*` (all columns)
+      return selectedColumns().isEmpty() || selectedColumns().contains(column);
+    }
+
+    /**
+     * Retrurn the list of columns whose values are used to distinguish one document from another.
+     */
+    private List<Column> docIdColumns(int keyDepth) {
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:549]: could this be `@Value.Lazy` as well?

File: restapi/src/test/java/io/stargate/db/datastore/AlphabeticalOrderPartitionKeyDecorator.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ivansenic) [reply restapi/src/test/java/io/stargate/db/datastore/AlphabeticalOrderPartitionKeyDecorator.java:28]: You see I missed that this is a test class, with the comment I get it immediately :smile:

File: restapi/src/test/java/io/stargate/web/docsapi/DocsApiTestSchemaProvider.java [hunk 24]
```diff
@@ -83,7 +84,7 @@ public DocsApiTestSchemaProvider(int maxDepth, String keyspaceName, String colle
           ImmutableColumn.builder()
               .name(QueryConstants.P_COLUMN_NAME.apply(i))
               .type(Column.Type.Text)
-              .kind(PartitionKey)
+              .kind(Clustering)
```
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/DocsApiTestSchemaProvider.java:87]: :warning: Wait is this correct change? You are changing the schema here?

File: restapi/src/test/java/io/stargate/web/docsapi/service/CombinedPagingStateTest.java (hunk 15 above)
Comment (ivansenic) [reply restapi/src/test/java/io/stargate/web/docsapi/service/CombinedPagingStateTest.java:78]: OK, resolve.

**Review by ivansenic — CHANGES_REQUESTED** [review]
As far as I am concerned we can merge this.. Awesome work @dimas-b, great testing.. There is only one change that worries me, thus set as *Request changes* until this is resolved...

File: restapi/src/test/java/io/stargate/web/docsapi/DocsApiTestSchemaProvider.java (hunk 24 above)
Comment (dimas-b) [reply restapi/src/test/java/io/stargate/web/docsapi/DocsApiTestSchemaProvider.java:87]: No, I believe old code was a mistake here. `pNNN` columns are clustering keys, not partition keys.

The reason why I changed it now, is that it breakes the new validation logic in `QueryExecutor`.

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 23 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:549]: sure, but I thought `@Value.Lazy` was not required in this case since this method is not a "getter" (it has an arg).

File: restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java (hunk 22 above)
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java:231]: good point. will fix :+1:

Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java]: fixed in ee5522221621015605a3fab3b196e37452a7843d

**Comment by dimas-b** [comment]
Rebased to `master`

File: restapi/src/test/java/io/stargate/web/docsapi/DocsApiTestSchemaProvider.java (hunk 24 above)
Comment (ivansenic) [reply restapi/src/test/java/io/stargate/web/docsapi/DocsApiTestSchemaProvider.java:87]: OK but once we redeploy this will only influence new tables right? What about the existing tables? If it brakes the validation logic will it not work in the existing tables?

Comment (dimas-b) [reply restapi/src/test/java/io/stargate/web/docsapi/DocsApiTestSchemaProvider.java:87]: Sorry, I'm not sure I understand... the `key` columns has always been the only partition key, the `pNNN` have always need clustering keys. This change affects only the test code. Production code is not affected and should work as before.

Comment (ivansenic) [reply restapi/src/test/java/io/stargate/web/docsapi/DocsApiTestSchemaProvider.java:87]: Fu*** I did not see this is a test class.. Sorry my bad :exploding_head:

**Review by ivansenic — APPROVED** [review]
Approved after all is resolved.
